src/inference/graph_inference.cpp

```diff
@@ -1101,10 +1166,63 @@ void GraphInference::SSVMLearn(
       VLOG(3) << a->GetLabelName(it->first.a_) << " " << a->GetLabelName(it->first.b_) << " " << a->GetLabelName(it->first.type_) << " " << it->second;
       auto features_it = features_.find(it->first);
       if (features_it != features_.end()) {
-        features_it->second.atomicAddRegularized(it->second, 0, svm_regularizer_);
+        features_it->second.atomicAddRegularized(it->second, 0, regularizer_);
+      }
+    }
+  }
+}
+
+void GraphInference::PLInit(double regularization, double margin, int beam_size, double pl_regularizer) {
+  regularizer_ = 1 / regularization;
+  for (auto it = features_.begin(); it != features_.end(); ++it) {
+    it->second.setValue(regularizer_ * 0.5);
+  }
+  beam_size_ = beam_size;
+  svm_margin_ = margin;
+  pl_regularizer_ = pl_regularizer;
+}
+
+void GraphInference::PLLearn(
+    const Nice2Query* query,
+    const Nice2Assignment* assignment,
+    double learning_rate,
+    int num_training_samples,
+    PrecisionStats* stats) {
+  const GraphNodeAssignment* a = static_cast<const GraphNodeAssignment*>(assignment);
+
+  GraphNodeAssignment new_assignment(*a);
+  new_assignment.SetUpEqualityPenalty(svm_margin_);
+  PerformAssignmentOptimization(&new_assignment);
+
+  UpdateStats((*a), new_assignment, stats, svm_margin_);
+
+  // Perform gradient descent
+  SimpleFeaturesMap affected_features;  // Gradient for each affected feature.
+  affected_features.set_empty_key(GraphFeature(-1, -1, -1));
+  affected_features.set_deleted_key(GraphFeature(-2, -2, -2));
+
+  for (int i = 0; i < a->assignments_.size(); i++) {
+    if (a->assignments_[i].must_infer) {
+      std::vector<int> candidates;
+      a->GetLabelCandidates(*this, i, &candidates, beam_size_);
+      double normalization_constant = a->GetNodeNormalizationConstantGivenNeighbours(*this, i, &candidates);
+      for (int label : candidates) {
+        double marginal_probability = a->GetNodeConditionalProbabilityGivenNeighbours(*this, i, normalization_constant, label);
+        a->GetNeghbouringAffectedFeature(&affected_features, i, label, -learning_rate * marginal_probability);
       }
     }
   }
+
+  a->GetAffectedFeatures(&affected_features, beam_size_ * learning_rate);
+  a->RegularizeAffectedFeature(&affected_features, num_training_samples * pow(pl_regularizer_, 2));
```

Comment:
This needs explanation. How is num_training_samples needed and what is this power? Did you want pow on 0.5 or 2 ?